### PR TITLE
Optimized Slasher Better DB Schema

### PR DESCRIPTION
**Pull request by rauljordan**
**What type of PR is this?**

> Optimization

**What does this PR do? Why is it needed?**

Currently, our optimized slasher has a not-well-optimized db schema, particularly for attestation records. We have a single bucket called `attestationRecordsBucket` which stores all attestations as follows:

```py
(target_epoch ++ validator_index) => (signing_root ++ compressed_ssz_encode(attestation))
```
The way we save to this bucket is as follows. Say we receive an attestation that looks like this:

```py
{
  AttestingIndices: [0, 1, 2, ..., 500],
  Data: 0xab
}
```

We perform the following routine for a given epoch:

```py
attesting_indices = attestation.attesting_indices
for validator in attesting_indices:
  put(validator++epoch, attestation)
```

This means if an attestation is 300 bytes, for example, we would be storing the same data duplicated 500 times! 150k bytes for this example sans-compression.

Instead, we propose the alternative schema:

**attestation_data_roots_bucket**
```py
(target_epoch ++ validator_index) => (signing_root ++ fastsum64(attestation.attesting_indices))
```

**attestation_records_bucket**
```py
(signing_root ++ fastsum64(attestation.attesting_indices) => compressed_ssz_encoded(attestation)
```
Then, our saving routine becomes:
```py
attesting_indices = attestation.attesting_indices
signing_root = signing_root(attestation)
attestation_records_bucket.put(signing_root++fastsum64(attesting_indices), encode(attestation))
for validator in attesting_indices:
  signing_roots_bucket.put(
    validator++epoch, signing_root++fastsum64(attesting_indices),
  )
```

If an attestation is 300 bytes, this is an 7.5x improvement to the current behavior in terms of disk use.

**Which issues(s) does this PR fix?**

Part of #8331

### CODE REVIEW

File: beacon-chain/db/kv/schema.go [hunk 1]
```diff
@@ -51,10 +51,11 @@ var (
 	newStateServiceCompatibleBucket = []byte("new-state-compatible")
 
 	// Slasher buckets.
-	attestedEpochsByValidator = []byte("attested-epochs-by-validator")
-	attestationRecordsBucket  = []byte("attestation-records")
-	proposalRecordsBucket     = []byte("proposal-records")
-	slasherChunksBucket       = []byte("slasher-chunks")
+	attestedEpochsByValidator  = []byte("attested-epochs-by-validator")
+	attestationRecordsBucket   = []byte("attestation-records")
+	attestationDataRootsBucket = []byte("attestation-data-roots")
```
Comment (rauljordan) [inline beacon-chain/db/kv/schema.go:56]: New one

File: beacon-chain/db/kv/slasher.go [hunk 2]
```diff
@@ -86,18 +87,23 @@ func (s *Store) CheckAttesterDoubleVotes(
...
+	err := s.db.View(func(tx *bolt.Tx) error {
+		signingRootsBkt := tx.Bucket(attestationDataRootsBucket)
+		attRecordsBkt := tx.Bucket(attestationRecordsBucket)
 		for _, att := range attestations {
 			encEpoch := encodeTargetEpoch(att.IndexedAttestation.Data.Target.Epoch, historyLength)
 			for _, valIdx := range att.IndexedAttestation.AttestingIndices {
 				encIdx := encodeValidatorIndex(types.ValidatorIndex(valIdx))
-				key := append(encEpoch, encIdx...)
-				encExistingAttRecord := bkt.Get(key)
-				if len(encExistingAttRecord) < 32 {
```
Comment (0xKiwi) [inline beacon-chain/db/kv/slasher.go:97]: Should we make these constants at the top of the file?

File: beacon-chain/db/kv/slasher.go [hunk 3]
```diff
@@ -155,22 +165,38 @@ func (s *Store) SaveAttestationRecordsForValidators(
...
 	return s.db.Update(func(tx *bolt.Tx) error {
-		bkt := tx.Bucket(attestationRecordsBucket)
+		attRecordsBkt := tx.Bucket(attestationRecordsBucket)
+		signingRootsBkt := tx.Bucket(attestationDataRootsBucket)
 		for i, att := range attestations {
+			// An attestation record key is comprised of the signing root (32 bytes)
+			// and a fastsum64 of the attesting indices (8 bytes). This is used
+			// to have a more optimal schema
+			attIndicesHash := hashutil.FastSum64(encodedIndices[i])
+			attRecordKey := append(att.SigningRoot[:], ssz.MarshalUint64(nil, attIndicesHash)...)
```
Comment (0xKiwi) [inline beacon-chain/db/kv/slasher.go:192]: Whats the `nil` here?

File: beacon-chain/db/kv/slasher.go [hunk 4]
```diff
@@ -340,17 +366,21 @@ func (s *Store) PruneAttestations(ctx context.Context, currentEpoch types.Epoch,
...
+		for k, v := c.Seek(epochEnc); k != nil; k, v = c.Prev() {
 			if !epochPrefixLessThan(k, epochEnc) {
 				continue
 			}
 			slasherAttestationsPrunedTotal.Inc()
-			if err := attBkt.Delete(k); err != nil {
+			if err := signingRootsBkt.Delete(k); err != nil {
+				return err
+			}
+			if err := attRecordsBkt.Delete(v); err != nil {
```
Comment (0xKiwi) [inline beacon-chain/db/kv/slasher.go:383]: We could be trying to delete the same key multiple times here, does that introduce additional overhead?

Comment (rauljordan) [reply beacon-chain/db/kv/slasher.go:383]: It's no cost. Delete is a no-op if the key doesn't exist, so we should be OK

**Review by 0xKiwi — APPROVED** [review]

**Pull request merged**